src/mesh/Triangle.cpp

```diff
@@ -24,54 +26,50 @@ Triangle::Triangle(
     Edge &edgeTwo,
     Edge &edgeThree,
     int   id)
-    : _edges({&edgeOne, &edgeTwo, &edgeThree}),
-      _id(id)
+    : _id(id)
 {
   PRECICE_ASSERT(edgeOne.getDimensions() == edgeTwo.getDimensions(),
                  edgeOne.getDimensions(), edgeTwo.getDimensions());
   PRECICE_ASSERT(edgeTwo.getDimensions() == edgeThree.getDimensions(),
                  edgeTwo.getDimensions(), edgeThree.getDimensions());
 
-  // Determine vertex map
-  Vertex &v0 = edge(0).vertex(0);
-  Vertex &v1 = edge(0).vertex(1);
-
-  if (&edge(1).vertex(0) == &v0) {
-    _vertexMap[0] = true;
-    _vertexMap[1] = false;
-  } else if (&edge(1).vertex(1) == &v0) {
-    _vertexMap[0] = true;
-    _vertexMap[1] = true;
-  } else if (&edge(1).vertex(0) == &v1) {
-    _vertexMap[0] = false;
-    _vertexMap[1] = false;
+  PRECICE_ASSERT(edgeOne.connectedTo(edgeTwo), "Edge one and two are not connected.");
+  PRECICE_ASSERT(edgeOne.connectedTo(edgeThree), "Edge one and three are not connected.");
+  PRECICE_ASSERT(edgeTwo.connectedTo(edgeThree), "Edge two and three are not connected.");
+
+  // Pick first 2 vertices from first edge
+  Vertex &v0 = edgeOne.vertex(0);
+  Vertex &v1 = edgeOne.vertex(1);
+
+  // Determine the third vertex using the second edge
+  Vertex *v2 = nullptr;
+  if ((v0 == edgeTwo.vertex(0)) || (v1 == edgeTwo.vertex(0))) {
+    v2 = &edgeTwo.vertex(1);
+  } else if ((v0 == edgeTwo.vertex(1)) || (v1 == edgeTwo.vertex(1))) {
+    v2 = &edgeTwo.vertex(0);
   } else {
-    PRECICE_ASSERT(&edge(1).vertex(1) == &v1);
-    _vertexMap[0] = false;
-    _vertexMap[1] = true;
+    PRECICE_UNREACHABLE("Edges don't form a triangle");
   }
 
-  if (_vertexMap[1] == 0) {
-    if (&edge(2).vertex(0) == &edge(1).vertex(1)) {
-      _vertexMap[2] = false;
-    } else {
-      PRECICE_ASSERT(&edge(2).vertex(1) == &edge(1).vertex(1));
-      _vertexMap[2] = true;
-    }
-  } else if (_vertexMap[1] == 1) {
-    if (&edge(2).vertex(0) == &edge(1).vertex(0)) {
-      _vertexMap[2] = false;
-    } else {
-      PRECICE_ASSERT(&edge(2).vertex(1) == &edge(1).vertex(0));
-      _vertexMap[2] = true;
-    }
-  }
+  _vertices = {&v0, &v1, v2};
+  std::sort(_vertices.begin(), _vertices.end(),
+            [](const Vertex *lhs, const Vertex *rhs) { return *lhs < *rhs; });
```

Comment:
Why do we need this sorting?

Reply:
By storing the vertices in id-ascending order, we can use lexicographical sort to establish an ordering.
The latter is a requirement for sorting.

Reply:
What, we need ordering for sorting? :laughing: Aren't these synonyms?
Why do we need to sort vertices in the first place? Not saying that we don't need to do it, but I did not understand why.

Reply:
An order establishes how things are ordered and allows comparing them. Sorting is an operation.

Establishing a weak partial order for vertices allows to sort vertices.
Sorting the vertices of connectivity primitives has two effects:
1. It maps all permutations to a single well defined case ( edges AB and BA both map to AB. triangles BCA, CAB etc map to ABC)
2. It allows to establish a weak partial order for connectivity primitives.
3. The combination of 1 and 2 allows to sort and remove redundant connectivity elements